fw/http.c

```diff
@@ -4824,164 +4786,279 @@ __tfw_h2_get_body_start(TfwHttpResp* resp)
+	TfwH2Ctx *ctx = tfw_h2_context(resp->req->conn);
+	TfwStream *stream = tfw_h2_find_stream(&ctx->sched, stream_id);
 
-		h_len -= max_sz;
-		frame_hdr.type = HTTP2_CONTINUATION;
-		mit->iter.frag = 0;
-		r = __tfw_h2_make_frames(h_len, data, sizeof(buf), iter, skew,
-					 max_sz, &frame_hdr_str, &frame_hdr,
-					 HTTP2_F_END_HEADERS);
-		if (unlikely(r))
-			return r;
+	if (!stream)
+		return -EPIPE;
+
+	r = tfw_h2_append_predefined_body(resp, stream_id, body);
+	if (unlikely(r))
+		return r;
+
+	tfw_h2_stream_init_for_xmit(stream, h_len, b_len);
+	return r;
+}
+
+static void
+tfw_h2_set_skb_flags(struct sk_buff *skb, struct sk_buff *nskb,
+		     TfwStream *stream)
+{
+	unsigned short flags = ss_skb_flags(skb) &
+		(SS_F_HTTT2_FRAME_HEADERS | SS_F_HTTT2_FRAME_DATA);
+
+	switch (flags) {
+	case SS_F_HTTT2_FRAME_HEADERS:
+		ss_skb_set_flags(nskb, SS_F_HTTT2_FRAME_HEADERS);
+		break;
+	case SS_F_HTTT2_FRAME_DATA:
+		ss_skb_set_flags(nskb, SS_F_HTTT2_FRAME_DATA);
+		break;
+	case SS_F_HTTT2_FRAME_HEADERS | SS_F_HTTT2_FRAME_DATA:
+		/* If headers located in old skb. */
+		if (skb->len - FRAME_HEADER_SIZE >= stream->xmit.h_len)
+			flags &= ~SS_F_HTTT2_FRAME_HEADERS;
+		ss_skb_set_flags(nskb, flags);
+		break;
+	default:
+		/* Should be checked early. */
+		BUG_ON(1);
 	}
 
-	if (local_response)
-		return 0;
+	ss_skb_set_priv_data(nskb, stream->id);
+	ss_skb_set_tls_type(nskb, ss_skb_tls_type(skb));
+}
 
-	/* Add more frame headers for DATA block. */
-	if (b_len > max_sz) {
-		unsigned long skew = 0;
+int
+tfw_h2_insert_frame_header(struct sock *sk,  struct sk_buff *skb,
+			   TfwStream *stream, unsigned int mss_now,
+			   TfwMsgIter *it, char *data,
+			   const TfwStr *frame_hdr_str,
+			   unsigned int *t_tz)
+{
+	struct sk_buff *nskb = skb->next;
+	unsigned len = skb->len, nskb_len = nskb->len;
+	unsigned int truesize = skb->truesize, n_truesize = nskb->truesize;
+	int delta, r;
+	long tmp_t_tz;
+	bool tcp_fragment = (skb->len != skb->data_len);
+
+	ss_skb_set_flags(skb, SS_F_IN_XMIT_QUEUE);
+	BUG_ON(skb->sk);
+	skb->sk = sk;
+	r = tfw_http_msg_insert(it, &data, frame_hdr_str);
+	skb->sk = NULL;
+	ss_skb_clear_flag(skb, SS_F_IN_XMIT_QUEUE);
+	if (unlikely(r))
+		return r;
 
-		/*
-		 * TODO: #498 and maybe #488 : iterate over the chunks only once
-		 * TODO: #1103 make only one memory allocation for hopefully
-		 * all the HTTP/2 frames headers instead of dealing with skb
-		 * framing with every small header
-		 */
-		if (test_bit(TFW_HTTP_B_CHUNKED, resp->flags)) {
-			data = h2_body->data;
-			iter->skb = h2_body->skb;
-		} else {
-			data = TFW_STR_CHUNK(&resp->body, 0)->data;
-			iter->skb = resp->body.skb;
+	delta = skb->len - len;
+	tcp_sk(sk)->write_seq += delta;
+	TCP_SKB_CB(skb)->end_seq += delta;
+	tmp_t_tz = (long int)skb->truesize - (long int)truesize;
+
+	if (skb->next != nskb) {
+		int old_factor;
+		u8 flags;
+
+		/* New skb was allocated and inserted after current. */
+		nskb = skb->next;
+
```

Comment:
All the code at the below, playing with TCP stuff - was it written from scratch or did you use some existing code as a reference? In the second case, please add a comment with the reference.

The TCP code is very sophisticated and we fixed a lot of bugs in the similar code in the TLS parts around `tfw_tls_tcp_propagate_dseq()`. During the TLS development I made some effort to separate the TCP logic (into `tfw_tls_tcp_propagate_dseq()`), but it isn't enough. This patch introduces even more TCP stuff (just compare `grep TCP_SKB_CB *.[ch]` with master). Is it possible to introduce `tcp.[ch]` files and move such logic there?

We benefit from the tight integration between the network layers, but it's still worth splitting the layers wherever it doesn't hurt performance. HTTP and TCP lives on different layers, there is TLS between them, and it's good to split them ad much as possible.

Reply:
This code was extracted from two functions `tcp_fragment` and `tco_fragment`